File: Graph_problems/FindAllPeopleWithSecret.py

# Problem link - https://leetcode.com/problems/find-all-people-with-secret/description/

from typing import List
from collections import defaultdict, deque

def findAllPeople(n: int, meetings: List[List[int]], firstPerson: int) -> List[int]:
    # step 1 - mark True for people who initially know the secret
    know_secret = [False] * n
    know_secret[0] = True
    know_secret[firstPerson] = True

    # step 2 - create time to meetings map (sorted dictionary)
    # where time should be sorted in ascending order
    time_meetings_map = defaultdict(list)
    for person1, person2, time in meetings:
        # Populate time-meetings map
        time_meetings_map[time].append((person1, person2))

    # step 3 - traverse the time-meeting map in sorted order
    for time in sorted(time_meetings_map.keys()):
        meets = time_meetings_map[time]

        # step 4 - build graph using adjacency list
        adj_list = defaultdict(list)
        queue = deque()
        visited = set()

        for p1, p2 in meets:
            adj_list[p1].append(p2)
            adj_list[p2].append(p1)

            # add person1 to queue and set if they know the secret
            if know_secret[p1] and p1 not in visited:
                queue.append(p1)
                visited.add(p1)

            # add person2 to queue and set if they know the secret
            if know_secret[p2] and p2 not in visited:
                queue.append(p2)
                visited.add(p2)

        # step 5 - spread secret by performing BFS traversal on graph
        while queue:
                person = queue.popleft()

                for next_person in adj_list[person]:
                    if not know_secret[next_person]:
                        know_secret[next_person] = True
                        queue.append(next_person)

    # step 6 - form result list and return it
    res = []
    for i in range(n):
        if know_secret[i]:
            res.append(i)
    return res
    
# n = 6
# meetings = [[1,2,5],[2,3,8],[1,5,10]]
# firstPerson = 1

# n = 4
# meetings = [[3,1,3],[1,2,2],[0,3,3]]
# firstPerson = 3

n = 5
meetings = [[3,4,2],[1,2,1],[2,3,1]]
firstPerson = 1
res = findAllPeople(n, meetings, firstPerson)
print(res)

File: Graph_problems/FindAllPeopleWithSecret.cpp

// Problem link - https://leetcode.com/problems/find-all-people-with-secret/

#include<bits/stdc++.h>
using namespace std;

typedef pair<int, int> P;

vector<int> findAllPeople(int n, vector<vector<int>>& meetings, int firstPerson) {
    // step 1 - mark true for person who initially knows secret
    vector<bool> knowSecret(n, false);
    knowSecret[0] = true;
    knowSecret[firstPerson] = true;

    // step 2 - create time to meetings map (ordered map)
    // where time should be sorted in ascending order
    map<int, vector<P>> timeMeetingsMap;
    for(vector<int>& meeting : meetings){
        int person1 = meeting[0];
        int person2 = meeting[1];
        int time = meeting[2];

        // populate time-meetings map
        timeMeetingsMap[time].push_back({person1, person2});
    }

    // step 3 - traverse in time-meeting map
    for(auto &it : timeMeetingsMap){
        // extract time and meeting
        int time = it.first;
        vector<P> meets = it.second;

        // step 4 - make graph using adjaceny list
        unordered_map<int, vector<int>> adjList;
        queue<int> qu;
        unordered_set<int> visited;
        for(auto& [p1, p2] : meets){
            adjList[p1].push_back(p2);
            adjList[p2].push_back(p1);

            // add person1 to queue and set
            if(knowSecret[p1] == true && visited.find(p1) == visited.end()){
                qu.push(p1);
                visited.insert(p1);
            }

            // add person2 to queue and set
            if(knowSecret[p2] == true && visited.find(p2) == visited.end()){
                qu.push(p2);
                visited.insert(p2);
            }
        }

        // step 5 - spread secret by performing BFS traversal on graph
        while (!qu.empty()){
            int person = qu.front();
            qu.pop();

            for(auto &nextperson : adjList[person]){
                if(knowSecret[nextperson] == false){
                    knowSecret[nextperson] = true;
                    qu.push(nextperson);
                }
            }
        }
    }

    // step 6 - form result, `res` and return it
    vector<int> res;
    for(int i=0; i<n; i++){
        if(knowSecret[i]){
            res.push_back(i);
        }
    }
    return res;
}

void printVector(vector<int>& vec){
    cout << "[";
    for (size_t i = 0; i < vec.size(); ++i) {
        cout<<vec[i];
        if (i < vec.size() - 1) {
            cout<<",";
        }
    }
    cout<< "]" <<endl;
}

int main(){
    // int n = 6;
    // vector<vector<int>> meetings = {{1,2,5},{2,3,8},{1,5,10}};
    // int firstPerson = 1;

    // int n = 4;
    // vector<vector<int>> meetings = {{3,1,3},{1,2,2},{0,3,3}};
    // int firstPerson = 3;

    int n = 5;
    vector<vector<int>> meetings = {{3,4,2},{1,2,1},{2,3,1}};
    int firstPerson = 1;
    vector<int> res = findAllPeople(n, meetings, firstPerson);
    printVector(res);
    return 0;
}